3) Feed the following paragraph into your favourite data analytics tool, and answer the following;

a. What is the probability of the word “data” occurring in each line ?

In [1]:
sentences = """As a term, data analytics predominantly refers to an assortment of applications, from basic business 
intelligence (BI), reporting and online analytical processing (OLAP) to various forms of advanced 
analytics. In that sense, it's similar in nature to business analytics, another umbrella term for 
approaches to analyzing data -- with the difference that the latter is oriented to business uses, while 
data analytics has a broader focus. The expansive view of the term isn't universal, though: In some 
cases, people use data analytics specifically to mean advanced analytics, treating BI as a separate 
category. Data analytics initiatives can help businesses increase revenues, improve operational 
efficiency, optimize marketing campaigns and customer service efforts, respond more quickly to 
emerging market trends and gain a competitive edge over rivals -- all with the ultimate goal of 
boosting business performance. Depending on the particular application, the data that's analyzed 
can consist of either historical records or new information that has been processed for real-time 
analytics uses. In addition, it can come from a mix of internal systems and external data sources. At 
a high level, data analytics methodologies include exploratory data analysis (EDA), which aims to find 
patterns and relationships in data, and confirmatory data analysis (CDA), which applies statistical 
techniques to determine whether hypotheses about a data set are true or false. EDA is often 
compared to detective work, while CDA is akin to the work of a judge or jury during a court trial -- a 
distinction first drawn by statistician John W. Tukey in his 1977 book Exploratory Data Analysis. Data 
analytics can also be separated into quantitative data analysis and qualitative data analysis. The 
former involves analysis of numerical data with quantifiable variables that can be compared or 
measured statistically. The qualitative approach is more interpretive -- it focuses on understanding 
the content of non-numerical data like text, images, audio and video, including common phrases, 
themes and points of view."""
print(sentences)

As a term, data analytics predominantly refers to an assortment of applications, from basic business 
intelligence (BI), reporting and online analytical processing (OLAP) to various forms of advanced 
analytics. In that sense, it's similar in nature to business analytics, another umbrella term for 
approaches to analyzing data -- with the difference that the latter is oriented to business uses, while 
data analytics has a broader focus. The expansive view of the term isn't universal, though: In some 
cases, people use data analytics specifically to mean advanced analytics, treating BI as a separate 
category. Data analytics initiatives can help businesses increase revenues, improve operational 
efficiency, optimize marketing campaigns and customer service efforts, respond more quickly to 
emerging market trends and gain a competitive edge over rivals -- all with the ultimate goal of 
boosting business performance. Depending on the particular application, the data that's analyzed 
can c

In [2]:
#Convert text to lowercase
sentences = sentences.lower()
print(sentences)

as a term, data analytics predominantly refers to an assortment of applications, from basic business 
intelligence (bi), reporting and online analytical processing (olap) to various forms of advanced 
analytics. in that sense, it's similar in nature to business analytics, another umbrella term for 
approaches to analyzing data -- with the difference that the latter is oriented to business uses, while 
data analytics has a broader focus. the expansive view of the term isn't universal, though: in some 
cases, people use data analytics specifically to mean advanced analytics, treating bi as a separate 
category. data analytics initiatives can help businesses increase revenues, improve operational 
efficiency, optimize marketing campaigns and customer service efforts, respond more quickly to 
emerging market trends and gain a competitive edge over rivals -- all with the ultimate goal of 
boosting business performance. depending on the particular application, the data that's analyzed 
can c

In [3]:
#Remove numbers
import re
sentences = re.sub(r'\d+', '', sentences)
print(sentences)

as a term, data analytics predominantly refers to an assortment of applications, from basic business 
intelligence (bi), reporting and online analytical processing (olap) to various forms of advanced 
analytics. in that sense, it's similar in nature to business analytics, another umbrella term for 
approaches to analyzing data -- with the difference that the latter is oriented to business uses, while 
data analytics has a broader focus. the expansive view of the term isn't universal, though: in some 
cases, people use data analytics specifically to mean advanced analytics, treating bi as a separate 
category. data analytics initiatives can help businesses increase revenues, improve operational 
efficiency, optimize marketing campaigns and customer service efforts, respond more quickly to 
emerging market trends and gain a competitive edge over rivals -- all with the ultimate goal of 
boosting business performance. depending on the particular application, the data that's analyzed 
can c

In [4]:
#Remove punctuation
import string
sentences = sentences.translate(str.maketrans('','', string.punctuation))
print(sentences)

as a term data analytics predominantly refers to an assortment of applications from basic business 
intelligence bi reporting and online analytical processing olap to various forms of advanced 
analytics in that sense its similar in nature to business analytics another umbrella term for 
approaches to analyzing data  with the difference that the latter is oriented to business uses while 
data analytics has a broader focus the expansive view of the term isnt universal though in some 
cases people use data analytics specifically to mean advanced analytics treating bi as a separate 
category data analytics initiatives can help businesses increase revenues improve operational 
efficiency optimize marketing campaigns and customer service efforts respond more quickly to 
emerging market trends and gain a competitive edge over rivals  all with the ultimate goal of 
boosting business performance depending on the particular application the data thats analyzed 
can consist of either historical r

In [5]:
#Remove whitespace
sentences = sentences.strip()
print(sentences)

as a term data analytics predominantly refers to an assortment of applications from basic business 
intelligence bi reporting and online analytical processing olap to various forms of advanced 
analytics in that sense its similar in nature to business analytics another umbrella term for 
approaches to analyzing data  with the difference that the latter is oriented to business uses while 
data analytics has a broader focus the expansive view of the term isnt universal though in some 
cases people use data analytics specifically to mean advanced analytics treating bi as a separate 
category data analytics initiatives can help businesses increase revenues improve operational 
efficiency optimize marketing campaigns and customer service efforts respond more quickly to 
emerging market trends and gain a competitive edge over rivals  all with the ultimate goal of 
boosting business performance depending on the particular application the data thats analyzed 
can consist of either historical r

In [7]:
#Split a string into a list
list_sentence = sentences.splitlines()
list_sentence

['as a term data analytics predominantly refers to an assortment of applications from basic business ',
 'intelligence bi reporting and online analytical processing olap to various forms of advanced ',
 'analytics in that sense its similar in nature to business analytics another umbrella term for ',
 'approaches to analyzing data  with the difference that the latter is oriented to business uses while ',
 'data analytics has a broader focus the expansive view of the term isnt universal though in some ',
 'cases people use data analytics specifically to mean advanced analytics treating bi as a separate ',
 'category data analytics initiatives can help businesses increase revenues improve operational ',
 'efficiency optimize marketing campaigns and customer service efforts respond more quickly to ',
 'emerging market trends and gain a competitive edge over rivals  all with the ultimate goal of ',
 'boosting business performance depending on the particular application the data thats analyz

In [8]:
#Number of list
number_of_list = len(list_sentence)
number_of_list

22

In [9]:
#Remove stop words and tokenization
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

nltk.download("stopwords")
nltk.download("punkt")

from nltk.tokenize import word_tokenize
tokens = word_tokenize(sentences)

stop_words = set(stopwords.words("english"))
result = [i for i in tokens if not i in stop_words]
#count word
import collections 
from collections import Counter

word_count_dict = Counter(result)
word_count_dict.most_common()

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


[('data', 18),
 ('analytics', 10),
 ('analysis', 6),
 ('business', 4),
 ('term', 3),
 ('bi', 2),
 ('advanced', 2),
 ('uses', 2),
 ('view', 2),
 ('exploratory', 2),
 ('eda', 2),
 ('cda', 2),
 ('compared', 2),
 ('work', 2),
 ('qualitative', 2),
 ('predominantly', 1),
 ('refers', 1),
 ('assortment', 1),
 ('applications', 1),
 ('basic', 1),
 ('intelligence', 1),
 ('reporting', 1),
 ('online', 1),
 ('analytical', 1),
 ('processing', 1),
 ('olap', 1),
 ('various', 1),
 ('forms', 1),
 ('sense', 1),
 ('similar', 1),
 ('nature', 1),
 ('another', 1),
 ('umbrella', 1),
 ('approaches', 1),
 ('analyzing', 1),
 ('difference', 1),
 ('latter', 1),
 ('oriented', 1),
 ('broader', 1),
 ('focus', 1),
 ('expansive', 1),
 ('isnt', 1),
 ('universal', 1),
 ('though', 1),
 ('cases', 1),
 ('people', 1),
 ('use', 1),
 ('specifically', 1),
 ('mean', 1),
 ('treating', 1),
 ('separate', 1),
 ('category', 1),
 ('initiatives', 1),
 ('help', 1),
 ('businesses', 1),
 ('increase', 1),
 ('revenues', 1),
 ('improve', 1),


In [10]:
def tokenize(string):
    return re.compile('\w+').findall(string)

from collections import Counter

def word_freq(string): 
    text = tokenize(string.lower())
    c = Counter(text)           # count the words
    d = Counter(''.join(text))  # count all letters
    return (dict(c),dict(d))    # return a tuple of counted words and letters


words, letters = word_freq(sentences) # count and get dicts with counts

#For the use of question (b)
def word_freq2(string): 
    text = tokenize(string.lower())
    c = Counter(text)           # count the words
    return (dict(c))    # return a tuple of counted words

words = word_freq2(sentences) # count and get dicts with counts

sumWords = sum(words.values())       # sum total words
sumLetters = sum(letters.values())   # sum total letters

# calculate or print probability of word
for w in words:
    print("Probability of '{}': {}".format(w,words[w]/sumWords))

Probability of 'as': 0.006369426751592357
Probability of 'a': 0.03184713375796178
Probability of 'term': 0.009554140127388535
Probability of 'data': 0.05732484076433121
Probability of 'analytics': 0.03184713375796178
Probability of 'predominantly': 0.0031847133757961785
Probability of 'refers': 0.0031847133757961785
Probability of 'to': 0.03503184713375796
Probability of 'an': 0.0031847133757961785
Probability of 'assortment': 0.0031847133757961785
Probability of 'of': 0.03184713375796178
Probability of 'applications': 0.0031847133757961785
Probability of 'from': 0.006369426751592357
Probability of 'basic': 0.0031847133757961785
Probability of 'business': 0.012738853503184714
Probability of 'intelligence': 0.0031847133757961785
Probability of 'bi': 0.006369426751592357
Probability of 'reporting': 0.0031847133757961785
Probability of 'and': 0.028662420382165606
Probability of 'online': 0.0031847133757961785
Probability of 'analytical': 0.0031847133757961785
Probability of 'processing': 

In [11]:
# calculate or print probability of letter
for l in letters:
    print("Probability of '{}': {}".format(l,letters[l]/sumLetters))

Probability of 'a': 0.12122969837587007
Probability of 's': 0.07540603248259861
Probability of 't': 0.0951276102088167
Probability of 'e': 0.10266821345707657
Probability of 'r': 0.0568445475638051
Probability of 'm': 0.02494199535962877
Probability of 'd': 0.0388631090487239
Probability of 'n': 0.07424593967517401
Probability of 'l': 0.04234338747099768
Probability of 'y': 0.0191415313225058
Probability of 'i': 0.08932714617169374
Probability of 'c': 0.03944315545243619
Probability of 'p': 0.023781902552204175
Probability of 'o': 0.06322505800464037
Probability of 'f': 0.019721577726218097
Probability of 'b': 0.011020881670533642
Probability of 'u': 0.02320185614849188
Probability of 'g': 0.01334106728538283
Probability of 'v': 0.013921113689095127
Probability of 'h': 0.027842227378190254
Probability of 'z': 0.0017401392111368909
Probability of 'w': 0.008700696055684454
Probability of 'x': 0.0034802784222737818
Probability of 'k': 0.005220417633410673
Probability of 'q': 0.00348027842

In [12]:
#Number of "data" word
count_data_word = words['data']
count_data_word

18

In [13]:
#Probability of the word “data” occurring in each line
print(count_data_word/number_of_list)

0.8181818181818182


Answer: Probability of the word “data” occurring in each line is 0.8182.

b. What is the distribution of distinct word counts across all the lines ?

In [14]:
#Find distinct/unique words
uniqueWords = [] 
for i in list_sentence:
      if not i in uniqueWords:
          uniqueWords.append(len(word_freq2(i)));  

In [15]:
#Print distinct/unique words
uniqueWords

[15,
 13,
 13,
 14,
 16,
 14,
 11,
 12,
 16,
 11,
 15,
 18,
 14,
 11,
 16,
 17,
 14,
 12,
 14,
 12,
 14,
 5]

Answer: The distribution of distinct word counts across all the lines are: <br>
1st line: 15 <br>
2nd line: 13 <br>
3rd line: 13<br>
4th line: 14<br>
5th line: 16<br>
6th line: 14<br>
7th line: 11<br>
8th line: 12<br>
9th line: 16<br>
10th line: 11<br>
11th line: 15<br>
12th line: 18<br>
13th line: 14<br>
14th line: 11<br>
15th line: 16<br>
16th line: 17<br>
17th line: 14<br>
18th line: 12<br>
19th line: 14<br>
20th line: 12<br>
21th line: 14<br>
22th line: 5<br>
 

c. What is the probability of the word “analytics” occurring after the word “data” ?

In [16]:
import nltk 
from nltk import word_tokenize
from nltk.util import ngrams

In [17]:
#Create bigrams 
bigrams = list(ngrams(tokens, 2))
bigrams

[('as', 'a'),
 ('a', 'term'),
 ('term', 'data'),
 ('data', 'analytics'),
 ('analytics', 'predominantly'),
 ('predominantly', 'refers'),
 ('refers', 'to'),
 ('to', 'an'),
 ('an', 'assortment'),
 ('assortment', 'of'),
 ('of', 'applications'),
 ('applications', 'from'),
 ('from', 'basic'),
 ('basic', 'business'),
 ('business', 'intelligence'),
 ('intelligence', 'bi'),
 ('bi', 'reporting'),
 ('reporting', 'and'),
 ('and', 'online'),
 ('online', 'analytical'),
 ('analytical', 'processing'),
 ('processing', 'olap'),
 ('olap', 'to'),
 ('to', 'various'),
 ('various', 'forms'),
 ('forms', 'of'),
 ('of', 'advanced'),
 ('advanced', 'analytics'),
 ('analytics', 'in'),
 ('in', 'that'),
 ('that', 'sense'),
 ('sense', 'its'),
 ('its', 'similar'),
 ('similar', 'in'),
 ('in', 'nature'),
 ('nature', 'to'),
 ('to', 'business'),
 ('business', 'analytics'),
 ('analytics', 'another'),
 ('another', 'umbrella'),
 ('umbrella', 'term'),
 ('term', 'for'),
 ('for', 'approaches'),
 ('approaches', 'to'),
 ('to', 'a

In [18]:
#Calculate frequency of bigrams
freq_bigram = nltk.FreqDist(bigrams)
freq_bigram

FreqDist({('data', 'analytics'): 6, ('data', 'analysis'): 5, ('as', 'a'): 2, ('advanced', 'analytics'): 2, ('to', 'business'): 2, ('data', 'with'): 2, ('with', 'the'): 2, ('exploratory', 'data'): 2, ('a', 'term'): 1, ('term', 'data'): 1, ...})

In [19]:
#List the most 10 common words
freq_bigram.most_common(10)

[(('data', 'analytics'), 6),
 (('data', 'analysis'), 5),
 (('as', 'a'), 2),
 (('advanced', 'analytics'), 2),
 (('to', 'business'), 2),
 (('data', 'with'), 2),
 (('with', 'the'), 2),
 (('exploratory', 'data'), 2),
 (('a', 'term'), 1),
 (('term', 'data'), 1)]

In [20]:
#The number of word with "analytics"
count_analytics_word = words['analytics']
count_analytics_word

10

In [21]:
#Probability of the word “analytics” occurring after the word “data"
probabilility_dataanalytics_word =  6 /count_analytics_word
print(probabilility_dataanalytics_word)

0.6


Answer: The probability of the word “analytics” occurring after the word “data" is 0.6.